design/ceph/object/store.md

```diff
@@ -351,3 +351,212 @@ spec:
+
+### Pool Placement targets and Storage Classes
+Object Storage API allows users to specify where bucket data will be stored during bucket creation. With `<LocationConstraint>` parameter in S3 API and `X-Storage-Policy` header in SWIFT. Similarly, users can override where object data will be stored by setting `X-Amz-Storage-Class` and `X-Object-Storage-Class` during object creation. 
+
+In Ceph these data locations represented by Pools. Placement targets control which Pools are associated with a particular bucket. A bucket’s placement target is selected on creation, and cannot be modified (See [Ceph doc](https://docs.ceph.com/en/latest/radosgw/placement/#pool-placement-and-storage-classes)). Placement target can also define custom Storage classes to override data pool where object will be stored.
+
+Ceph administrator creates set of `index_pool`, `data_pool`, and `data_extra_pool` per each target placement. Select arbitrary names for target placements. For example `hot` and `cold`:
+```yaml
+# Bucket Data pool for hot storage
+apiVersion: ceph.rook.io/v1
+kind: CephBlockPool
+metadata:
+  name: hot.rgw.buckets.data
+  namespace: rook-ceph 
+spec:
+  ...
+---
+# Bucket index pool for hot storage
+apiVersion: ceph.rook.io/v1
+kind: CephBlockPool
+metadata:
+  name: hot.rgw.buckets.index
+  namespace: rook-ceph
+spec:
+  ...
+---
+# Bucket data extra pool for hot storage
+apiVersion: ceph.rook.io/v1
+kind: CephBlockPool
+metadata:
+  name: hot.rgw.buckets.non-ec
+  namespace: rook-ceph
+spec:
+  ...
+---
+# Bucket Data pool for cold storage
+apiVersion: ceph.rook.io/v1
+kind: CephBlockPool
+metadata:
+  name: cold.rgw.buckets.data
+  namespace: rook-ceph 
+spec:
+  ...
+---
+# Bucket index pool for cold storage
+apiVersion: ceph.rook.io/v1
+kind: CephBlockPool
+metadata:
+  name: cold.rgw.buckets.index
+  namespace: rook-ceph
+spec:
+  ...
+---
+# Bucket data extra pool for cold storage
+apiVersion: ceph.rook.io/v1
+kind: CephBlockPool
+metadata:
+  name: cold.rgw.buckets.non-ec
+  namespace: rook-ceph 
+spec:
+  ...
+```
+If needed, create additional pool to use it in custom StorageClass, for example `GLACIER`:
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephBlockPool
+metadata:
+  name: glacier.rgw.buckets.data
+  namespace: rook-ceph 
+spec:
+  ...
+```
+Then update `CephObjectStore` CRD with target placements and StorageClasses definitions:
+```yaml
+spec:
+  targetPlacements:                         # [optional] list of target placements
+  # target placement named hot:
+  - name: hot                               # [required] should be unique
+    dataPool: hot.rgw.buckets.data          # [required] default data pool, will be used with Ceph default STANDARD StorageClass, can be overridden with custom StorageClasses
+    indexPool: hot.rgw.buckets.index        # [required] 
+    dataExtraPool: hot.rgw.buckets.non-ec   # [required] 
+    default: false                          # true|false only one target must be set as default
```

Comment:
I'd like to discuss the `default` option more.

IMO, it will be a pain for us to enforce a list of structs where only a single default is true using CRD validation rules, and I'm not sure it will be a very big benefit. I'd like to consider having the top-level pool selection (i.e., the existing configs that are outside of `targetPlacement`) always be the default, and documenting that fact for users.

I'm curious about @anthonyeleven 's thoughts on this as well though.

Reply:
Probably we can have something like this:
```yaml
spec:
  targetPlacements: [optional]
    default: cold [optional]
    placements:
      hot:
        dataPool: hot.rgw.buckets.data
        indexPool: hot.rgw.buckets.index
        dataExtraPool: hot.rgw.buckets.non-ec
      cold:
        dataPool: cold.rgw.buckets.data
        indexPool: cold.rgw.buckets.index
        dataExtraPool: cold.rgw.buckets.non-ec
```
Another question: should we remove old default pools `<CephObjectStore name>.rgw.buckets...` if user sets `targetPlacements.default`?

Reply:
I do generally tend to think that explicit is better than implicit. 

As you suggest, if the target placements are all explicitly set, we don't need to use the top-level settings. And in to help with explicit-ness, I think that might be preferable.

If we go this route, which I am in favor of, let us be sure to (at minimum) add verification in the code to ensure `targetPlacements` and existing pool configs are not both set. Rook should error with a helpful message if both are set. I would also like to see at least some basic Validating Admission Policy CEL rules as well if possible.

Reply:
Are you saying the CephObjectStore metadata and data pool settings should be mutually exclusive with the `targetPlacements`? Rather than mutually exclusive, I was thinking of these as additional placements. I'd say we still need the pools that come with the objects store, so we can ensure all the core object store pools are created (IIRC there are about 6 of them). The target placements only require three pools. So we could use the pools named `<CephObjectStore name>.rgw.buckets...` as one of the target placements, and treat the `targetPlacements` as additional placements to add. 

For the default placement, sounds good to have the `default` field at the top level as suggested. This way, the `<CephObjectStore name>` could also be set, even though it's not in the `targetPlacements` list. 

That said, perhaps it would be helpful to see some full example manifests to illustrate both approaches to compare which one is simpler from the user perspective. My impression is that removing the pools will just have extra complications for implementation.